### PR TITLE
Solves  #2846 Refactor run_opts into a @dataclass

### PR DESCRIPTION
**Pull request by nouranali**
## What does this PR do?

Fixes #<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

**Comment by nouranali** [comment]
@pplantinga

**Comment by pplantinga** [comment]
Hi @nouranali - thanks for your contribution. This is a step in the right direction, but I think there's more that's needed here. Just to warn you, the `core.py` and `interfaces.py` are important parts of the code and will need a lot of testing before this sort of PR can be merged.

So to start with, I think `Brain` and `Pretrained` should accept a `RunOpts` object as an argument, rather than a dict. In addition, the `parse_arguments()` function is closely tied to the RunOpts, and should maybe just be a method of the class or at least in the same file, and should probably just return a `RunOpts` object (e.g. `RunOpts.from_command_line_args()` `@classmethod` or some such thing).  Finally, there is some argument validation that is done in `Brain` that could maybe move to the RunOpts class. Use your best judgement -- this part is probably the tricky part of the PR, as we should focus on making the Brain class an easy to understand basic training loop implementation, while no surprising details should get hidden if possible. Good luck!

**Comment by nouranali** [comment]
> Hi @nouranali - thanks for your contribution. This is a step in the right direction, but I think there's more that's needed here. Just to warn you, the `core.py` and `interfaces.py` are important parts of the code and will need a lot of testing before this sort of PR can be merged.
> 
> So to start with, I think `Brain` and `Pretrained` should accept a `RunOpts` object as an argument, rather than a dict. In addition, the `parse_arguments()` function is closely tied to the RunOpts, and should maybe just be a method of the class or at least in the same file, and should probably just return a `RunOpts` object (e.g. `RunOpts.from_command_line_args()` `@classmethod` or some such thing). Finally, there is some argument validation that is done in `Brain` that could maybe move to the RunOpts class. Use your best judgement -- this part is probably the tricky part of the PR, as we should focus on making the Brain class an easy to understand basic training loop implementation, while no surprising details should get hidden if possible. Good luck!

Great, will start working on your comments, how can I start testing along the way?

**Comment by nouranali** [comment]
@pplantinga Can u check the new updates?

**Comment by nouranali** [comment]
@pplantinga Can u check now?

**Comment by pplantinga** [comment]
I don't have a lot of time to review this rn, but you can check it yourself by running a template, e.g. [`templates/speech_recognition/ASR`](https://github.com/speechbrain/speechbrain/tree/develop/templates/speech_recognition/ASR)

**Comment by pplantinga** [comment]
My only comments right now are: the `from_command_line_args()` method should be a `@classmethod`. And ideally the hparams file and yaml_overrides should become members of the RunOpts. I think the name `RunOptDefaults` should be changed to something like `RunOptions`.

https://docs.python.org/3/library/functions.html#classmethod

**Comment by nouranali** [comment]
> My only comments right now are: the `from_command_line_args()` method should be a `@classmethod`. And ideally the hparams file and yaml_overrides should become members of the RunOpts. I think the name `RunOptDefaults` should be changed to something like `RunOptions`.
> 
> https://docs.python.org/3/library/functions.html#classmethod

Done

**Comment by nouranali** [comment]
@mravanelli

**Comment by nouranali** [comment]
> @mravanelli

Is there a way for the checks to be automatically run like docs??

**Comment by nouranali** [comment]
@mravanelli

File: speechbrain/core.py [hunk 1]
```diff
@@ -658,8 +322,9 @@ def __init__(  # noqa: C901
         self.optimizers_dict = None
         self.opt_class = opt_class
         self.checkpointer = checkpointer
+        _,self.run_opt_defaults,_ = RunOptions.from_command_line_args()
 
-        for arg, default in run_opt_defaults.items():
+        for arg, default in self.run_opt_defaults.items():
             if run_opts is not None and arg in run_opts:
                 if hparams is not None and arg in hparams:
                     logger.info(
```
Comment (pplantinga) [inline speechbrain/core.py:330]: The defaults are all handled by the class now. So this shouldn't have to call `from_command_line_args()` at all, which can be done in the recipes. Instead, the class should just take a `RunOptions` object.

File: speechbrain/utils/run_opts.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        # Checking that DataParallel use the right number of GPU
+        if run_opts.data_parallel_backend and torch.cuda.device_count() == 0:
+            raise ValueError("You must have at least 1 GPU.")
+
+        # force device arg to be the same as local_rank from torchrun
+        local_rank = os.environ.get("LOCAL_RANK")
+        if local_rank is not None and "cuda" in run_opts.device:
+            run_opts.devide = run_opts.devide[:-1] + str(local_rank)
+
+        return param_file, run_opts.as_dict(), overrides
```
Comment (pplantinga) [inline speechbrain/utils/run_opts.py:332]: This should return a `RunOptions` instance for the second object. While it would be cleaner to return *only* a run options object, I think we'll have to keep the return signature for compatibility, so we don't have to update every single recipe.

**Review by pplantinga — COMMENTED** [review]
At this point I'm unsure if this PR will need to change every recipe or not -- ideally we could get away without having to do this, but we'll see. Have you tried running a template yet?

**Comment by pplantinga** [comment]
I ran the checks and the linter failed. You can run it yourself, even automatically on every commit, following instructions here:
https://github.com/speechbrain/speechbrain/blob/develop/docs/devtools.md

**Comment by pplantinga** [comment]
This used to touch only a few files but now touches over 500... did you run a different version of black on the whole repo?

**Comment by nouranali** [comment]
> This used to touch only a few files but now touches over 500... did you run a different version of black on the whole repo?

What version do u use? and what files do u run it on?

**Comment by pplantinga** [comment]
If you install the `precommit` hook, the linters will run on all the files changed as part of the commit. The versions are in the file https://github.com/speechbrain/speechbrain/blob/develop/lint-requirements.txt

File: speechbrain/core.py [hunk 3]
```diff
@@ -510,89 +177,28 @@ class and override any methods for which the default behavior does not
...
         trainable parameters, and will have ``train()``/``eval()`` called on them.
-    opt_class : torch.optim class
+    opt_class : Optional[Type[torch.optim]]
         A torch optimizer constructor that takes only the list of
         parameters (e.g. a lambda or partial function definition). By default,
         this will be passed all modules in ``modules`` at the
         beginning of the ``fit()`` method. This behavior can be changed
         by overriding the ``configure_optimizers()`` method.
-    hparams : dict
+    hparams : Optional[dict]
```
Comment (Adel-Moumen) [inline speechbrain/core.py:189]: I am wondering if we should maybe log this to the user i.e. saying that no opt+class / hparams are provided etc to make sure everything is transparent to the user

**Comment by TParcollet** [comment]
@pplantinga / @Adel-Moumen if agreement from both of you, please merge!
